File: backend/pkg/aiproxy/url.go
package aiproxy

import (
	"net"
	"net/url"
	"strings"
)

// normalizeLLMBaseURL validates outbound LLM API base URLs (SSRF-safe).
// HTTPS: allowed for public hosts. HTTP: only local/docker allowlist (Ollama etc.).
func normalizeLLMBaseURL(raw string) (string, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", ErrInvalidBaseURL
	}
	u, err := url.Parse(raw)
	if err != nil {
		return "", ErrInvalidBaseURL
	}
	host := strings.ToLower(u.Hostname())
	switch u.Scheme {
	case "https":
		if isBlockedHost(host) {
			return "", ErrInvalidBaseURL
		}
	case "http":
		if !isAllowedLocalLLMHost(host) {
			return "", ErrInvalidBaseURL
		}
	default:
		return "", ErrInvalidBaseURL
	}
	return ensureOpenAIV1Suffix(raw), nil
}

func ensureOpenAIV1Suffix(raw string) string {
	raw = strings.TrimSuffix(strings.TrimSpace(raw), "/")
	if strings.HasSuffix(raw, "/v1") {
		return raw
	}
	return raw + "/v1"
}

func isAllowedLocalLLMHost(host string) bool {
	allowed := map[string]bool{
		"127.0.0.1":            true,
		"localhost":            true,
		"::1":                  true,
		"ollama":               true,
		"host.docker.internal": true,
	}
	return allowed[host]
}

func isBlockedHost(host string) bool {
	if host == "" {
		return true
	}
	if isAllowedLocalLLMHost(host) {
		// Cloud keys should not target loopback over HTTPS in production proxy.
		return false
	}
	ip := net.ParseIP(host)
	if ip == nil {
		return false
	}
	return ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast()
}

File: py-backend/app/url.py
import ipaddress
from urllib.parse import urlparse

ALLOWED_LOCAL_HOSTS = {
    "127.0.0.1",
    "localhost",
    "::1",
    "ollama",
    "host.docker.internal",
}


class InvalidBaseURL(Exception):
    pass


def normalize_llm_base_url(raw: str) -> str:
    raw = (raw or "").strip()
    if not raw:
        raise InvalidBaseURL("ai base url not allowed")

    parsed = urlparse(raw)
    host = (parsed.hostname or "").lower()
    scheme = (parsed.scheme or "").lower()

    if scheme == "https":
        if _is_blocked_host(host):
            raise InvalidBaseURL("ai base url not allowed")
    elif scheme == "http":
        if host not in ALLOWED_LOCAL_HOSTS:
            raise InvalidBaseURL("ai base url not allowed")
    else:
        raise InvalidBaseURL("ai base url not allowed")

    return _ensure_openai_v1_suffix(raw)


def _ensure_openai_v1_suffix(raw: str) -> str:
    raw = raw.rstrip("/")
    if raw.endswith("/v1"):
        return raw
    return raw + "/v1"


def _is_blocked_host(host: str) -> bool:
    if not host:
        return True
    if host in ALLOWED_LOCAL_HOSTS:
        return False
    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        return False
    return (
        ip.is_loopback
        or ip.is_private
        or ip.is_link_local
        or ip.is_multicast
    )
